api/envoy/extensions/transport_sockets/tls/v3/common.proto

```diff
@@ -235,6 +249,18 @@ message TlsCertificate {
   // error.
   PrivateKeyProvider private_key_provider = 6;
 
+  // [#not-implemented-hide:]
+  // This provides a list of BoringSSL private key method provider. Envoy will find out
+  // an available private key method provider. It may fallback to BoringSSL default implementation
+  // when there is no available one. All the private key provider will share the same private key
+  // in the :ref:`private_key <envoy_v3_api_field_extensions.transport_sockets.tls.v3.TlsCertificate.private_key>` field,
+  // so the :ref:`private_key <envoy_v3_api_field_extensions.transport_sockets.tls.v3.TlsCertificate.private_key>` field
+  // must be specified when the `proviate_key_provider_list` field is used. The old :ref:`private_key_provider
+  // <envoy_v3_api_field_extensions.transport_sockets.tls.v3.TlsCertificate.private_key_provider>` field will be
+  // deprecated. If both :ref:`private_key_provider <envoy_v3_api_field_extensions.transport_sockets.tls.v3.TlsCertificate.private_key_provider>`
+  // and `private_key_provider_list` are provided, the config will be rejected.
```

Comment:
Instead of saying that the config will be rejected, why not say that if `private_key_provider_list` is present, `private_key_provider` is ignored?  That way, there's a clean upgrade path for control planes to support both old and new clients.

Reply:
yea,, got it, I will change. This reminder me the `optional HTTP filter` case. The control-plane can just send single config, then old and new envoy can work with it.